proposals/expand-ref.md

```diff
@@ -0,0 +1,294 @@
+
+```csharp
+void Swap(ref scoped Span<int> p1, ref scoped Span<int> p2)
+{
+    // Error: can't assign current parameter 2 to current parameter 1
+    p2 = p1;
+
+    // Error: can't assign current parameter 1 to current parameter 2
+    p1 = p2;
+
+    // Okay: as current parameter 1 and 2 can be converted to current method
+    scoped Span<int> local1 = p1; 
+    scoped Span<int> local2 = p2; 
+
+    // Okay: however the safe-to-escape here is current parameter X, not 
+    // current method so this could cause a bit of confusion later on
+    Span<int> local3 = p1; 
+    Span<int> local4 = p2; 
+
+    // Okay: the safe-to-escape of the value is inferred in this case as it is 
+    // done for ref locals today.
+    ref Span<int> refLocal1 = ref p1;
+    ref Span<int> refLocal2 = ref p2;
+}
+```
+
+A `ref scoped` parameter is also implicitly `scoped ref`. That means neither the value nor it's `ref` can be returned from the method. Both `ref` and `in` parameters can have their values modified with `scoped`. An `out` parameter cannot have its value modified with `scoped` as such a declaration is non-sensical. 
+
+```csharp
+void M(
+    ref scoped Span<int> p1,    // Okay
+    in scoped Span<int> p2,     // Okay
+    out scoped Span<int> p2,    // Error
+)
+```
+
+The [method arguments must match](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/low-level-struct-improvements.md#rules-method-arguments-must-match) rules will be updated to take `ref scoped` into account. Values passed to such parameters do not need to be considered when calculating the return scopes.
+
+Detailed notes:
+- A `ref scoped` parameter is implicitly `scoped ref`
+- A `out scoped` parameter declaration is an error
+
+### ref field to ref struct
+The language will allow for `ref struct` to appear as `ref scoped` fields. This `scoped` will serve to ensure the values cannot be escaped outside the containing instance but can be read and manipulated within it. 
+
+```csharp
+ref struct Deserializer
+{
+    ref scoped Utf8JsonReader reader;
+
+    ReadOnlySpan<byte> M1()
+    {
+        // okay: implicitly scoped to current method
+        var span = reader.ValueSpan; 
+
+        // okay
+        reader.Skip();
+
+        // Error: can't escape the ref data the ref scoped field refers to
+        return reader.ValueSpan;
+    }
+}
+```
+
+This is accomplished by giving every `ref scoped` field a new two new escape scopes named _current field N_ and _current ref field N_ where _N_ is the numeric order of the field. For example the first field has a _safe-to-escape_ of _current field 1_ and a _ref-safe-to-escape_ of _current ref field N_. Both escape scopes can be converted to _current method_, and _current field N_ can be converted to _current ref field N_, but no other defined relationships exist. That serves to restrict their usage to the current method where the containing value is used. This escape scope applies to both 
```

Comment:
I think that all lifetimes `$current_field_N` and `$current_ref_field_N` must also be convertible to lifetime `$this`. In the absence of this constraint, it would be possible for the `this` variable to hold references to fields with narrower lifetimes.

Reply:
Or, depending on how you look at it, it simply wouldn't be possible to declare any 'ref scoped' fields in ref structs, because the field's lifetime would not be convertible to the lifetime of the containing 'this' variable.

Reply:
I disagree. Logically yes I agree that the lifetimes must be convertible to `$this`. The moment you allow that though in all contexts, suddenly it's not `scoped` anymore, it's returnable 

```csharp
struct S<$this, $refField1, $field1>
  where $refField : $this
  where $field1: $this, $refField1 {
  ref<$refField1> Span<$field1, byte> span;
}

Span<$cm> M<$cm, $l1, $l2>(S<$cm, $l1, $l2> s) => 
  // works because $l2 is convertible to $cm
  s.span;
}
```

The way this works is that at construction time the rules ensure that the lifetimes of both the `ref` and the value are convertible to `$this`. That just falls out from our existing method args rules. Given that the fields lifetimes are invariant there is no need to keep that relationship alive for correctness after that. The only thing we have to worry about is how they are interpreted in the context of a method. 

Note: if these were `ref unscoped` fields then you are 100% correct we'd put that relationship in place at the decl of the type.